File: lintcode/892.alien-dictionary.cpp
#include <algorithm>
#include <queue>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
public:
  /**
   * @param words: a list of words
   * @return: a string which is correct order
   */
  string alienOrder(vector<string> &words) {
    string res = "";
    if (words.size() <= 1) {
      return res;
    }
    unordered_map<char, unordered_set<char>> edges;
    unordered_set<char> nodes;
    unordered_map<char, int> indegrees;

    for (auto const &word : words) {
      nodes.insert(cbegin(word), cend(word));
    }

    for (char node : nodes) {
      edges[node];
      indegrees[node];
    }

    for (int i = 0; i < words.size() - 1; ++i) {
      string wordA = words[i];
      string wordB = words[i + 1];
      int n = wordA.size();
      int m = wordB.size();

      int j = 0;
      for (; j < min(n, m); ++j) {
        if (wordA[j] != wordB[j]) {
          edges[wordA[j]].insert(wordB[j]);
          break;
        }
      }

      if (j == min(n, m) && n > m) {
        return "";
      }
    }
    for (const auto &kv : edges) {
      for (char n : kv.second) {
        indegrees[n] += 1;
      }
    }
    priority_queue<char, vector<char>, greater<char>> q;
    for (const auto &kv : indegrees) {
      if (kv.second == 0) {
        q.push(kv.first);
      }
    }

    while (!q.empty()) {
      char front = q.top();
      q.pop();
      res += front;
      for (char n : edges[front]) {
        indegrees[n] -= 1;
        if (indegrees[n] == 0) {
          q.push(n);
        }
      }
    }
    return res;
  }
};

File: lintcode/892.alien-dictionary.py
class Solution:
    """
    @param words: a list of words
    @return: a string which is correct order
    """

    def alienOrder(self, words):
        nodes = set()
        edges = {}
        indegrees = {}
        for word in words:
            for ch in word:
                nodes.add(ch)

        for node in nodes:
            edges[node] = set()
            indegrees[node] = 0

        for i in range(len(words) - 1):
            m = len(words[i])
            n = len(words[i + 1])

            mn = min(m, n)

            idx = 0

            while idx < mn:
                ch1 = words[i][idx]
                ch2 = words[i + 1][idx]
                if ch1 != ch2:
                    edges[ch1].add(ch2)
                    break
                idx += 1

            if idx == mn and m > n:
                return ""

        for node in nodes:
            for nei in edges[node]:
                indegrees[nei] += 1

        import heapq
        q = []

        for node in nodes:
            print(f'node: {node}, degrees: {indegrees[node]}')
            if indegrees[node] == 0:
                heapq.heappush(q, node)

        res = ""
        while len(q) > 0:
            node = heapq.heappop(q)
            res += node
            for nei in edges[node]:
                indegrees[nei] -= 1
                if indegrees[nei] == 0:
                    heapq.heappush(q, nei)
        if len(res) == len(nodes):
            return res
        else:
            return ""


if __name__ == '__main__':
    #print(Solution().alienOrder(["wrt","wrf","er","ett","rftt"]))
    #print(Solution().alienOrder([  "z", "x" ]))
    print(Solution().alienOrder(["abc", "bcd", "qwert", "ab"]))
